haystack/nodes/retriever/_openai_encoder.py

```diff
@@ -0,0 +1,144 @@
+import json
+import logging
+import sys
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Union
+
+import numpy as np
+import requests
+from tqdm.auto import tqdm
+
+from haystack.errors import OpenAIError, OpenAIRateLimitError
+from haystack.nodes.retriever._base_encoder import _BaseEmbeddingEncoder
+from haystack.schema import Document
+from haystack.utils.reflection import retry_with_exponential_backoff
+from haystack.utils.import_utils import _optional_component_not_installed
+
+if TYPE_CHECKING:
+    from haystack.nodes.retriever import EmbeddingRetriever
+
+USE_TIKTOKEN = False
+if sys.version_info >= (3, 8):
+    USE_TIKTOKEN = True
+
+if USE_TIKTOKEN:
+    try:
+        import tiktoken
+    except (ImportError, ModuleNotFoundError) as ie:
+        _optional_component_not_installed(__name__, "openai", ie)
+else:
+    from transformers import GPT2TokenizerFast, PreTrainedTokenizerFast
+
+
+logger = logging.getLogger(__name__)
+
+
+class _OpenAIEmbeddingEncoder(_BaseEmbeddingEncoder):
+    def __init__(self, retriever: "EmbeddingRetriever"):
+        # See https://beta.openai.com/docs/guides/embeddings for more details
+        self.url = "https://api.openai.com/v1/embeddings"
+        self.api_key = retriever.api_key
+        self.batch_size = min(64, retriever.batch_size)
+        self.progress_bar = retriever.progress_bar
+        model_class: str = next(
+            (m for m in ["ada", "babbage", "davinci", "curie"] if m in retriever.embedding_model), "babbage"
+        )
+
+        tokenizer = self._setup_encoding_models(model_class, retriever.embedding_model, retriever.max_seq_len)
+
+        if USE_TIKTOKEN:
+            logger.debug("Using tiktoken %s tokenizer", tokenizer)
+            self._tk_tokenizer: tiktoken.Encoding = tiktoken.get_encoding(tokenizer)
+        else:
+            logger.debug("Using GPT2TokenizerFast tokenizer")
+            self._hf_tokenizer: PreTrainedTokenizerFast = GPT2TokenizerFast.from_pretrained(tokenizer)
+
+    def _setup_encoding_models(self, model_class: str, model_name: str, max_seq_len: int):
+        """
+        Setup the encoding models for the retriever.
+        """
+
+        tokenizer_name = "gpt2"
+        # new generation of embedding models (December 2022), we need to specify the full name
+        if "text-embedding" in model_name:
```

Comment:
I'm wondering if it would be worth just checking for the `-002` tag here to detect the new generation of models, since that's how they are denoting generations according to this doc [page](https://beta.openai.com/docs/guides/embeddings/embedding-models). What do you think?

Reply:
Yes, I didn't update this specific flow and left it unmodified. Indeed, `-002` suffix is just for the embedding models.